core/bindings.rs

```diff
@@ -0,0 +1,88 @@
+use crate::isolate::Buf;
+use crate::libdeno::PinnedBuf;
+use futures::Future;
+use std::fmt;
+use std::io;
+
+pub type BindingResult<T> = std::result::Result<T, BindingError>;
+
+#[derive(Debug)]
+pub struct BindingError {
+  repr: Repr,
+}
+
+#[derive(Debug)]
+enum Repr {
+  Simple(String),
+  IoErr(io::Error),
+}
+
+pub fn new_binding_error(msg: String) -> BindingError {
+  BindingError {
+    repr: Repr::Simple(msg),
+  }
+}
+
+impl fmt::Display for BindingError {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    match self.repr {
+      Repr::Simple(ref err_str) => f.pad(err_str),
+      Repr::IoErr(ref err) => err.fmt(f),
+    }
+  }
+}
+
+impl std::error::Error for BindingError {
+  fn description(&self) -> &str {
+    match self.repr {
+      Repr::Simple(ref msg) => msg.as_str(),
+      Repr::IoErr(ref err) => err.description(),
+    }
+  }
+
+  fn cause(&self) -> Option<&dyn std::error::Error> {
+    match self.repr {
+      Repr::Simple(ref _msg) => None,
+      Repr::IoErr(ref err) => Some(err),
+    }
+  }
+}
+
+impl From<io::Error> for BindingError {
+  #[inline]
+  fn from(err: io::Error) -> Self {
+    Self {
+      repr: Repr::IoErr(err),
+    }
+  }
+}
+
+pub type BindingOpAsyncFuture =
+  Box<dyn Future<Item = Buf, Error = BindingError> + Send>;
+
+pub enum BindingOpSyncOrAsync {
+  Sync(Buf),
+  Async(BindingOpAsyncFuture),
+}
+
+pub type BindingOpResult = BindingResult<BindingOpSyncOrAsync>;
```

Comment:
This is confusing, so many errors. Would prefer if you just used Op<BindingError> instead.

Reply:
It makes for cleaner implementation though. We can use the try(`?`) operator. Still going to simplify a little here though:
``` rust
/// Base result type for a plugin op represents either a Sync or Async value
pub type PluginOp = Op<PluginError>;

/// Complete return type for a plugin op including Sync errors
pub type PluginOpResult = PluginResult<PluginOp>;

/// Funciton type for plugin ops
pub type PluginOpDispatchFn =
  fn(is_sync: bool, data: &[u8], zeroCopy: Option<PinnedBuf>)
    -> PluginOpResult;
```